src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -542,33 +544,31 @@ public long Address
             }
         }
 
-        internal bool Equals(object comparandObj, bool compareScopeId)
+        /// <summary>Compares two IP addresses.</summary>
+        public override bool Equals(object comparand)
         {
-            IPAddress comparand = comparandObj as IPAddress;
+            return comparand is IPAddress address && Equals(address);
+        }
 
-            if (comparand == null)
-            {
-                return false;
-            }
+        internal bool Equals(IPAddress comparand)
+        {
+            Debug.Assert(comparand != null);
 
             // Compare families before address representations
             if (AddressFamily != comparand.AddressFamily)
             {
                 return false;
             }
+
             if (IsIPv6)
             {
-                // For IPv6 addresses, we must compare the full 128-bit representation.
-                for (int i = 0; i < NumberOfLabels; i++)
-                {
-                    if (comparand._numbers[i] != _numbers[i])
-                    {
-                        return false;
-                    }
-                }
-
-                // The scope IDs must also match
-                return comparand.PrivateScopeId == PrivateScopeId || !compareScopeId;
+                // For IPv6 addresses, we must compare the full 128-bit representation and the scope IDs.
+                ReadOnlySpan<byte> thisNumbers = MemoryMarshal.AsBytes<ushort>(_numbers);
+                ReadOnlySpan<byte> comparandNumbers = MemoryMarshal.AsBytes<ushort>(comparand._numbers);
+                return
+                    MemoryMarshal.Read<ulong>(thisNumbers) == MemoryMarshal.Read<ulong>(comparandNumbers) &&
+                    MemoryMarshal.Read<ulong>(thisNumbers.Slice(sizeof(ulong))) == MemoryMarshal.Read<ulong>(comparandNumbers.Slice(sizeof(ulong))) &&
```

Comment:
I wonder if it can be boosted with SSE 🙂 
```csharp
if (Sse41.IsSupported)
{ 
    fixed (ushort* num1ptr = _numbers)
    fixed (ushort* num2ptr = comparand._numbers)
    {
        return Sse41.TestAllOnes( // or Sse2.MoveMask(...) == 0xFFFF which one is faster
            Sse2.CompareEqual(Sse2.LoadVector128(num1ptr), Sse2.LoadVector128(num2ptr)));
    }
}
```